src/fhirdefs/load.ts

```diff
@@ -1,29 +1,113 @@
-// TODO: Load from package files instead of these static folders.
 import { FHIRDefinitions } from './FHIRDefinitions';
-import fs from 'fs';
+import { PackageLoadError, DevPackageLoadError } from '../errors';
+import fs from 'fs-extra';
+import path from 'path';
+import os from 'os';
+import tar from 'tar';
+import rp from 'request-promise-native';
+import temp from 'temp';
 
-const _cache: Map<string, FHIRDefinitions> = new Map();
+/**
+ * Loads a dependency from user FHIR cache or from online
+ * @param {string} packageName - The name of the package to load
+ * @param {string} version - The version of the package to load
+ * @param {FHIRDefinitions} FHIRDefs - The FHIRDefinitions to load the dependencies into
+ * @param {string} cachePath - The path to load the package into
+ * @returns {Promise<FHIRDefinitions>} the loaded FHIRDefs
+ * @throws {PackageLoadError} when the desired package can't be loaded
+ */
+export async function loadDependency(
+  packageName: string,
+  version: string,
+  FHIRDefs: FHIRDefinitions,
+  cachePath: string = path.join(os.homedir(), '.fhir', 'packages')
+): Promise<FHIRDefinitions> {
+  const fullPackageName = `${packageName}#${version}`;
+  const loadPath = path.join(cachePath, fullPackageName, 'package');
+  let loadedPackage: string;
+  if (version != 'current') {
+    loadedPackage = loadFromPath(loadPath, fullPackageName, FHIRDefs);
+  }
+  if (!loadedPackage) {
+    let packageUrl: string;
+    if (version === 'dev') {
+      // Dev packages must be present in local FHIR cache
+      throw new DevPackageLoadError(fullPackageName);
+    } else if (version === 'current') {
+      // Current packages need to be loaded using build.fhir.org
+      const baseUrl = 'http://build.fhir.org/ig';
+      const res = await rp.get({ uri: `${baseUrl}/qas.json`, encoding: null });
+      const decoded = String.fromCharCode(...new Uint16Array(res));
+      const qaJSON = JSON.parse(decoded);
```

Comment:
Since the response body is expected to be JSON, it should be possible make the request with the option `json: true`. This will automatically parse the response body as JSON, so you won't need to do that manually. Additionally, it would be useful to include some error handling in case the package list is unavailable or broken.

Reply:
Ah true, that is neater, good catch. I'll also add some error handling.